File: shipley_checks/src/Stage_4/metrics/m7_pollinator_support.R
#!/usr/bin/env Rscript
#
# METRIC 7: POLLINATOR SUPPORT (SHARED POLLINATORS)
#
# ============================================================================
# ECOLOGICAL RATIONALE
# ============================================================================
#
# Shared pollinators create mutualistic networks that benefit both plants
# and pollinators:
#
# 1. RESOURCE CONTINUITY:
#    - Multiple plants provide sequential/overlapping flowering
#    - Reduces seasonal gaps in nectar/pollen availability
#    - Supports stable pollinator populations
#
# 2. POLLINATOR FIDELITY:
#    - Generalist pollinators visit multiple plant species in sequence
#    - Increases pollen transfer efficiency
#    - Higher visitation rates benefit plant reproduction
#
# 3. NETWORK RESILIENCE:
#    - If one plant species fails (drought, disease), pollinators have
#      alternative resources
#    - Redundancy protects against pollinator population crashes
#
# KEY CONCEPT: Quadratic Benefit
# The value of shared pollinators increases NON-LINEARLY with plant overlap:
# - 2 plants sharing a pollinator: Moderate benefit
# - 5 plants sharing a pollinator: HIGH benefit (pollinator visits all frequently)
# - Quadratic weighting: (count / n_plants)²
#
# This reflects:
# - Increased pollinator retention (more resources = longer stays)
# - Network effects (pollinator movements between plants)
# - Flowering synchrony benefits
#
# ============================================================================
# CALCULATION STEPS
# ============================================================================
#
# STEP 1: Count shared pollinators across guild members
#   - For each pollinator taxon, count how many plants host it
#   - Only pollinators shared by ≥2 plants contribute to score
#   - Uses ONLY strict pollinators (NOT flower_visitors - contaminated)
#
# STEP 2: Score with quadratic weighting
#   - For each shared pollinator:
#     * overlap_ratio = count / n_plants
#     * contribution = overlap_ratio²
#   - Sum across all shared pollinators
#   - Interpretation:
#     * 2/7 plants (29% overlap) → 0.08 contribution
#     * 5/7 plants (71% overlap) → 0.51 contribution (6× higher!)
#     * 7/7 plants (100% overlap) → 1.00 contribution (12× higher!)
#
# STEP 3: Percentile normalization
#   - Uses Köppen tier-stratified calibration for metric 'p6'
#   - HIGH p7_raw → HIGH percentile (good pollinator support)
#   - Note: No final inversion needed (high percentile = high score)
#
# ============================================================================
# DATA SOURCES
# ============================================================================
#
# organisms_df (organism_profiles_pure_r.csv):
# - Column: plant_wfo_id, pollinators (ONLY - NOT flower_visitors)
# - Contains pipe-separated GBIF animal taxon IDs
# - Derived from GloBI interaction network (interactionTypeName == 'pollinates')
# - Includes only strict pollinators (bees, specialized flies, hummingbirds)
# - DATA QUALITY NOTE: flower_visitors column is contaminated with:
#   * Herbivore pests (mites, caterpillars) feeding on flowers
#   * Pathogenic fungi growing on flowers
#   * Other non-pollinator organisms found on flowers
#
# ============================================================================
# PARITY REQUIREMENTS
# ============================================================================
#
# To maintain 100% parity with Python scorer:
# 1. Must use same count_shared_organisms logic (≥2 plants threshold)
# 2. Must use same quadratic weighting formula: (count / n_plants)²
# 3. Must use ONLY pollinators column (NOT flower_visitors - contaminated)
# 4. Must use same Köppen tier calibration file for 'p6' metric
#
# Python implementation: guild_scorer_v3.py lines 1645-1674
# NOTE: Python version also updated to use only 'pollinators' column
#
# ============================================================================


#' Calculate M7: Pollinator Support
#'
#' Scores shared pollinator networks using quadratic weighting to reflect
#' non-linear benefits of high-overlap pollinator communities.
#'
#' @param plant_ids Character vector of WFO taxon IDs for the guild
#' @param guild_plants Data frame with plant metadata (not used in M7)
#' @param organisms_df Data frame with plant-organism associations
#' @param count_shared_organisms_fn Function to count shared organisms
#' @param percentile_normalize_fn Function to convert raw score to percentile
#'
#' @return List with three elements:
#'   - raw: Quadratic-weighted pollinator overlap score
#'   - norm: Percentile score (0-100, HIGH = GOOD)
#'   - details: List with n_shared_pollinators and top 5 pollinator names
#'
calculate_m7_pollinator_support <- function(plant_ids,
                                            guild_plants,
                                            organisms_df,
                                            count_shared_organisms_fn,
                                            percentile_normalize_fn) {

  n_plants <- length(plant_ids)

  # Count shared pollinators (pollinators hosted by ≥2 plants)
  # Uses ONLY strict pollinators (NOT flower_visitors - contaminated with herbivores/fungi)
  shared_pollinators <- count_shared_organisms_fn(
    organisms_df, plant_ids,
    'pollinators'  # ONLY verified pollinators
  )

  # Score with QUADRATIC weighting
  # Reflects non-linear benefits of high-overlap pollinator communities
  p7_raw <- 0
  for (org_name in names(shared_pollinators)) {
    count <- shared_pollinators[[org_name]]
    if (count >= 2) {
      overlap_ratio <- count / n_plants
      p7_raw <- p7_raw + overlap_ratio ^ 2  # QUADRATIC benefit
    }
  }

  # Percentile normalize
  m7_norm <- percentile_normalize_fn(p7_raw, 'p6')

  # -------------------------------------------------------------------------
  # STEP 4: Build pollinator counts and categorization (for network profile)
  # -------------------------------------------------------------------------
  # Build pollinator_counts: pollinator_name → number of plants it visits
  # Also categorize each pollinator by taxonomic group

  categorize_pollinator <- function(name) {
    # Convert to lowercase for case-insensitive matching
    name_lower <- tolower(name)

    # Honey Bees (Apis) - most specific first
    if (grepl("\\bapis\\b", name_lower)) {
      return("Honey Bees")
    }
    # Bumblebees (Bombus)
    if (grepl("bombus", name_lower)) {
      return("Bumblebees")
    }
    # Hover Flies (Syrphidae) - before general "fly"
    if (grepl("syrph|episyrphus|eristalis|eupeodes|melanostoma|platycheirus|sphaerophoria|cheilosia", name_lower)) {
      return("Hover Flies")
    }
    # Mosquitoes (Culicidae) - before general "fly"
    if (grepl("aedes|culex|anopheles|culiseta|mosquito", name_lower)) {
      return("Mosquitoes")
    }
    # Muscid Flies (Muscidae/Anthomyiidae) - before general "fly"
    if (grepl("anthomyia|\\bmusca\\b|fannia|phaonia|delia|drymeia|muscidae", name_lower)) {
      return("Muscid Flies")
    }
    # Solitary Bees (after Apis/Bombus, before general "bee")
    if (grepl("andrena|lasioglossum|halictus|osmia|megachile|ceratina|xylocopa|anthophora|anthidium|colletes|nomada|agapostemon|amegilla|trigona|melipona|eulaema|epicharis|augochlora|chelostoma|tetralonia|bee", name_lower)) {
      return("Solitary Bees")
    }
    # Other Flies (catch remaining Diptera)
    if (grepl("fly|empis|calliphora|scathophaga|drosophila|bibio|diptera|rhamphomyia", name_lower)) {
      return("Other Flies")
    }
    # Pollen Beetles (before general "beetle")
    if (grepl("meligethes|brassicogethes|oedemera", name_lower)) {
      return("Pollen Beetles")
    }
    # Other Beetles
    if (grepl("beetle|cetonia|trichius|anaspis|coleoptera", name_lower)) {
      return("Other Beetles")
    }
    # Butterflies (Lepidoptera - Rhopalocera)
    if (grepl("papilio|pieris|vanessa|danaus|colias|lycaena|polyommatus|aglais|coenonympha|erebia|gonepteryx|anthocharis|maniola|butterfly", name_lower)) {
      return("Butterflies")
    }
    # Moths (Lepidoptera - Heterocera)
    if (grepl("moth|sphinx|manduca|hyles|macroglossum", name_lower)) {
      return("Moths")
    }
    # Wasps (Hymenoptera - non-Apoidea)
    if (grepl("wasp|vespula|vespa|polistes|dolichovespula", name_lower)) {
      return("Wasps")
    }
    # Birds
    if (grepl("bird|hummingbird|trochilidae|amazilia|phaethornis|coereba|anthracothorax|\\baves\\b", name_lower)) {
      return("Birds")
    }
    # Bats
    if (grepl("bat|chiroptera|pteropus|artibeus", name_lower)) {
      return("Bats")
    }

    return("Other")
  }

  # Build pollinator counts from organisms_df
  guild_organisms <- organisms_df %>% dplyr::filter(plant_wfo_id %in% plant_ids)
  pollinator_counts <- list()
  pollinator_category_map <- list()

  for (i in seq_len(nrow(guild_organisms))) {
    row <- guild_organisms[i, ]

    # Aggregate from ONLY pollinators column (NOT flower_visitors - contaminated)
    col_val <- row[['pollinators']]
    if (is.list(col_val)) col_val <- col_val[[1]]
    if (!is.null(col_val) && length(col_val) > 0) {
      for (pollinator in col_val) {
        if (is.null(pollinator_counts[[pollinator]])) {
          pollinator_counts[[pollinator]] <- 0
          pollinator_category_map[[pollinator]] <- categorize_pollinator(pollinator)
        }
        pollinator_counts[[pollinator]] <- pollinator_counts[[pollinator]] + 1
      }
    }
  }

  list(
    raw = p7_raw,
    norm = m7_norm,
    pollinator_counts = pollinator_counts,
    pollinator_category_map = pollinator_category_map,
    details = list(
      n_shared_pollinators = length(shared_pollinators),
      pollinators = names(shared_pollinators)[1:min(5, length(shared_pollinators))]
    )
  )
}

File: shipley_checks/src/Stage_4/guild_scorer_rust/src/metrics/m7_pollinator_support.rs
//! METRIC 7: POLLINATOR SUPPORT (SHARED POLLINATORS)
//!
//! **PHASE 3 OPTIMIZATION**: Pre-filtered LazyFrame with column projection
//!
//! Scores shared pollinator networks using quadratic weighting to reflect
//! non-linear benefits of high-overlap pollinator communities.
//!
//! **Memory optimization**:
//!   - Old: Receives full organisms_df (11,711 rows), filters inside count_shared_organisms
//!   - New: Receives pre-filtered organisms_lazy (7 rows), selects only 2 needed columns
//!   - Reuses same organisms_lazy as M3 (no redundant filtering!)
//!
//! **Columns needed** (M7 selects these 2):
//!   1. plant_wfo_id - Plant identification (for filtering)
//!   2. pollinators - Pipe-separated pollinator IDs (strict pollinators only)
//!
//! **Data Quality Note**:
//!   - Uses ONLY "pollinators" column (GloBI interactionTypeName == 'pollinates')
//!   - Does NOT use "flower_visitors" (contaminated with herbivores, fungi, etc.)
//!   - flower_visitors includes mites, caterpillars, and pathogenic fungi
//!
//! R reference: shipley_checks/src/Stage_4/metrics/m7_pollinator_support.R

use polars::prelude::*;
use anyhow::Result;
use rustc_hash::FxHashMap;
use crate::utils::{Calibration, percentile_normalize, count_shared_organisms, materialize_with_columns, filter_to_guild};

/// Column requirements for M7 calculation (from organisms parquet)
pub const REQUIRED_ORGANISM_COLS: &[&str] = &[
    "plant_wfo_id",
    "pollinators",  // ONLY pollinators - flower_visitors contaminated with herbivores/fungi
];

/// Result of M7 calculation
#[derive(Debug)]
pub struct M7Result {
    /// Coverage percentage: % of plants with ≥1 documented pollinator (0-100)
    pub raw: f64,
    /// Percentile score (0-100, HIGH = GOOD)
    pub norm: f64,
    /// Quadratic-weighted pollinator overlap score (OLD metric, kept for reporting)
    pub quadratic_score: f64,
    /// Number of shared pollinators
    pub n_shared_pollinators: usize,
    /// Number of plants with ≥1 documented pollinator
    pub plants_with_pollinators: usize,
    /// Total plants in guild
    pub total_plants: usize,
    /// Map of pollinator_name → plant_count for detailed analysis
    pub pollinator_counts: FxHashMap<String, usize>,
}

/// Calculate M7: Pollinator Support
///
/// **PHASE 3 OPTIMIZATION**: Reuses organisms LazyFrame from scorer, filters after column projection
///
/// R reference: m7_pollinator_support.R::calculate_m7_pollinator_support
pub fn calculate_m7(
    plant_ids: &[String],        // Guild plant IDs for filtering
    organisms_lazy: &LazyFrame,  // Schema-only scan (from scorer, reused from M3!)
    calibration: &Calibration,
) -> Result<M7Result> {
    // STEP 1: Materialize organisms columns and filter to guild
    let organisms_df = materialize_with_columns(
        organisms_lazy,
        REQUIRED_ORGANISM_COLS,
        "M7 organisms",
    )?;

    let guild_organisms = filter_to_guild(&organisms_df, plant_ids, "plant_wfo_id", "M7")?;

    let n_plants = guild_organisms.height();

    // Get plant IDs from the filtered DataFrame (now local to this function)
    let guild_plant_ids: Vec<String> = guild_organisms
        .column("plant_wfo_id")?
        .str()?
        .into_iter()
        .filter_map(|opt| opt.map(|s| s.to_string()))
        .collect();

    // Count shared pollinators (pollinators hosted by ≥2 plants)
    // Uses ONLY strict pollinators (NOT flower_visitors - contaminated with herbivores/fungi)
    let shared_pollinators = count_shared_organisms(
        &guild_organisms,
        &guild_plant_ids,
        &["pollinators"],  // Match R: ONLY verified pollinators
    )?;

    // Calculate OLD quadratic-weighted score (kept for reporting)
    // Reflects non-linear benefits of high-overlap pollinator communities
    // Formula: Sum of (Number of plants visited by Pollinator X / Total Plants)^2 for all Pollinator Xs
    let mut quadratic_score = 0.0;
    for (_org_name, count) in &shared_pollinators {
        if *count >= 2 {
            let overlap_ratio = *count as f64 / n_plants as f64;
            quadratic_score += overlap_ratio.powi(2); // QUADRATIC benefit
        }
    }

    // Calculate coverage percentage: % of plants with ≥1 documented pollinator
    let plants_with_pollinators = count_plants_with_pollinators(&guild_organisms, &guild_plant_ids)?;
    let coverage_pct = if n_plants > 0 {
        (plants_with_pollinators as f64 / n_plants as f64) * 100.0
    } else {
        0.0
    };

    // Percentile normalize (using coverage %)
    let m7_norm = percentile_normalize(coverage_pct, "p6", calibration, false)?;

    Ok(M7Result {
        raw: coverage_pct,  // Coverage % (0-100)
        norm: m7_norm,      // Percentile
        quadratic_score,    // OLD metric (kept for reporting)
        n_shared_pollinators: shared_pollinators.len(),
        plants_with_pollinators,
        total_plants: n_plants,
        pollinator_counts: shared_pollinators,
    })
}

/// Count how many plants have at least one documented pollinator
fn count_plants_with_pollinators(
    organisms_df: &DataFrame,
    plant_ids: &[String],
) -> Result<usize> {
    use rustc_hash::FxHashSet;

    let plant_id_set: FxHashSet<&String> = plant_ids.iter().collect();
    let plant_col = organisms_df.column("plant_wfo_id")?.str()?;

    let mut count = 0;
    for idx in 0..organisms_df.height() {
        if let Some(plant_id) = plant_col.get(idx) {
            if !plant_id_set.contains(&plant_id.to_string()) {
                continue; // Not in guild
            }

            // Check if this plant has any pollinators
            if let Ok(pol_col) = organisms_df.column("pollinators") {
                // Phase 0-4 parquets use Arrow list columns
                if let Ok(list_col) = pol_col.list() {
                    if let Some(list_series) = list_col.get_as_series(idx) {
                        if list_series.len() > 0 {
                            count += 1;
                        }
                    }
                } else if let Ok(str_col) = pol_col.str() {
                    // Fallback: pipe-separated strings (legacy format)
                    if let Some(value) = str_col.get(idx) {
                        if !value.is_empty() && value.split('|').any(|s| !s.is_empty()) {
                            count += 1;
                        }
                    }
                }
            }
        }
    }

    Ok(count)
}

#[cfg(test)]
mod tests {
    use approx::assert_relative_eq;

    #[test]
    fn test_quadratic_weighting() {
        let n_plants = 7.0_f64;

        // 2 plants → overlap = 2/7 ≈ 0.286 → squared ≈ 0.082
        let overlap = 2.0_f64 / n_plants;
        let contribution = overlap.powi(2);
        assert_relative_eq!(contribution, 0.082, epsilon = 0.001);

        // 5 plants → overlap = 5/7 ≈ 0.714 → squared ≈ 0.51
        let overlap = 5.0_f64 / n_plants;
        let contribution = overlap.powi(2);
        assert_relative_eq!(contribution, 0.51, epsilon = 0.01);

        // 7 plants → overlap = 7/7 = 1.0 → squared = 1.0
        let overlap = 7.0_f64 / n_plants;
        let contribution = overlap.powi(2);
        assert_relative_eq!(contribution, 1.0, epsilon = 0.0001);
    }

    #[test]
    fn test_multiple_pollinators() {
        let n_plants = 5.0_f64;

        let mut p7_raw = 0.0_f64;

        // Pollinator 1: 3 plants
        let overlap = 3.0_f64 / n_plants;
        p7_raw += overlap.powi(2);

        // Pollinator 2: 4 plants
        let overlap = 4.0_f64 / n_plants;
        p7_raw += overlap.powi(2);

        // Total: (3/5)² + (4/5)² = 0.36 + 0.64 = 1.0
        assert_relative_eq!(p7_raw, 1.0, epsilon = 0.0001);
    }
}
